File: dimos/manipulation/visualization/viser/runtime.py
# Copyright 2025-2026 Dimensional Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

from typing import TYPE_CHECKING, Protocol, cast
import webbrowser

from dimos.manipulation.visualization.viser.config import ViserVisualizationConfig

if TYPE_CHECKING:
    import viser
    from viser.extras import ViserUrdf as ViserUrdfType

VISER_INSTALL_HINT = (
    "Viser manipulation visualization requires Viser. Install it with: uv sync --extra manipulation"
)
VISER_URDF_INSTALL_HINT = (
    "Viser URDF support requires yourdfpy. Install it with: uv sync --extra manipulation"
)


class _Stoppable(Protocol):
    def stop(self) -> None: ...


def import_viser() -> type[viser.ViserServer]:
    """Import Viser with a feature-specific install hint."""
    try:
        from viser import ViserServer
    except ModuleNotFoundError as e:
        if e.name != "viser":
            raise
        raise ModuleNotFoundError(VISER_INSTALL_HINT) from e
    return ViserServer


def import_viser_urdf() -> type[ViserUrdfType]:
    """Import ViserUrdf with a feature-specific install hint."""
    try:
        from viser.extras import ViserUrdf
    except ModuleNotFoundError as e:
        if e.name not in {"viser", "viser.extras", "yourdfpy"}:
            raise
        raise ModuleNotFoundError(VISER_URDF_INSTALL_HINT) from e
    except ImportError as e:
        if "ViserUrdf" not in str(e):
            raise
        raise ModuleNotFoundError(VISER_URDF_INSTALL_HINT) from e
    return ViserUrdf


class ViserRuntime:
    """Owns the Viser server lifecycle."""

    def __init__(self, config: ViserVisualizationConfig) -> None:
        self.config = config
        self.server: viser.ViserServer | None = None

    @property
    def url(self) -> str | None:
        if self.server is None:
            return None
        return f"http://{self.config.host}:{self.config.port}"

    def start(self) -> viser.ViserServer:
        if self.server is None:
            ViserServer = import_viser()
            self.server = ViserServer(host=self.config.host, port=self.config.port)
            if self.config.open_browser and self.url:
                webbrowser.open_new_tab(self.url)
        return self.server

    def close(self) -> None:
        server = self.server
        self.server = None
        if server is not None and hasattr(server, "stop"):
            cast("_Stoppable", server).stop()

File: dimos/evals/agents/lib/runtime.js
// Copyright 2026 Dimensional Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { readFileSync, writeFileSync } from "node:fs";

/** @type {{provider: string, base_url: string, key_env: string, allowed_tools: string[] | null, max_output_tokens: number | null}} */
const config = JSON.parse(
  readFileSync(new URL("./runtime.json", import.meta.url), "utf8"),
);
const ready = new URL("./runtime-ready.json", import.meta.url);

/** @param {import("@earendil-works/pi-coding-agent").ExtensionAPI} pi */
export default function (pi) {
  pi.registerProvider(config.provider, {
    baseUrl: config.base_url,
    apiKey: "$" + config.key_env,
  });
  const apply = () => {
    const allowed = config.allowed_tools;
    const names = new Set(pi.getAllTools().map((tool) => tool.name));
    const unknown = allowed?.filter((name) => !names.has(name)) ?? [];
    if (allowed !== null) pi.setActiveTools(unknown.length ? [] : allowed);
    writeFileSync(
      ready,
      JSON.stringify({ tools: pi.getActiveTools(), unknown }),
    );
  };
  pi.on("session_start", apply);
  pi.on("before_agent_start", apply);
  pi.on("tool_call", (event) => {
    if (
      config.allowed_tools !== null &&
      !config.allowed_tools.includes(event.toolName)
    )
      return { block: true, reason: "Tool excluded by eval allowed_tools" };
  });
  pi.on("before_provider_request", (event) => {
    const cap = config.max_output_tokens;
    if (cap === null) return;
    if (!event.payload || typeof event.payload !== "object")
      throw new Error("Invalid provider payload");
    const payload = /** @type {Record<string, unknown>} */ (event.payload);
    const field = [
      "max_output_tokens",
      "max_tokens",
      "max_completion_tokens",
    ].find((name) => name in payload);
    if (!field)
      throw new Error("Provider does not expose an output token limit");
    const requested = payload[field];
    return {
      ...payload,
      [field]: typeof requested === "number" ? Math.min(requested, cap) : cap,
    };
  });
}
